### ISSUE TITLE
AMP with multi GPU

### ISSUE THREAD

**Issue by Smirkey** [facebookresearch/detectron2#2244]
Hello !

You recently added pytorch AMP in detectron2, i wanted to give it a shot. But when I tried training on multi gpu I saw no memory improvement.
When looking at the commits that adds amp support I saw that the main thing done is
`with autocast():
    out=model(data)
    loss=loss_fn(out)
`
The following from the pytorch blog is not taken into account

![image](https://user-images.githubusercontent.com/25055427/98579877-73184380-22bf-11eb-9ee4-a30417669fe7.png)

There's no autocast decorator upside forwards functions in meta-archs classes or sub-parts of backbones, that maybe explains the fact that it had no effect on multi gpu

Could it be an oversight ? 

Thanks in advance for your response
D2 is the best :)

**Comment by ppwwyyxx** [comment]
We do not use "DataParallel in a single process" so the above screenshot is not relevant.

PyTorch's AMP is not expected to save memory significantly.

**Comment by Smirkey** [comment]
https://pytorch.org/blog/accelerating-training-on-nvidia-gpus-with-pytorch-automatic-mixed-precision/

- Shorter training time;
**- Lower memory requirements, enabling larger batch sizes, larger models, or larger inputs.**

I did not made up the info about reduced memory usage, it's written on pytorch and nvidia blogs, as well as in the original paper.
D2 does not use "DataParallel in a single process" but it does use DistributedDataParallel, so the problem remains the same.
The AMPTrainer will work with multiple GPUs but won't apply AMP to the model, which will be trained normally.

**Comment by Smirkey** [comment]
Btw
> PyTorch's AMP is not expected to save memory significantly.

from pytorch blog
- Shorter training time;
- **Lower memory requirements, enabling larger batch sizes, larger models, or larger inputs.**

https://pytorch.org/blog/accelerating-training-on-nvidia-gpus-with-pytorch-automatic-mixed-precision/

have a nice day :)

**Comment by ppwwyyxx** [comment]
PyTorch's AMP is not expected to save memory significantly, for models in detectron2. It does save memory a little bit. For example, we have observed 100MB saving for RetinaNet and 200MB saving for Mask R-CNN.

**Comment by ppwwyyxx** [comment]
84a09834f6d838534951907fd9ef90fec73614d2 could reduce memory usage by a another few hundred MB for certain models.

**Comment by Smirkey** [comment]
> 200MB saving for Mask R-CNN.

I observe the same thing. I tested other backbones that are not supported by D2 atm, ones that use depthwise separable conv, the memory reduction at training time is much higher.

Thank you very much for your quick and very clear answers !
